### PR TITLE
Support array literal functions

### PR DESCRIPTION
**Pull request by xiangfu0**
Support array literal in thrift and wire for both v1/v2 engine.
Note that, the scalar functions returning `string[]` should be registered in `TransformFunctionType`

### CODE REVIEW

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java [hunk 1]
```diff
@@ -156,7 +156,12 @@ public enum ObjectType {
     FloatArrayList(44),
     StringArrayList(45),
     UltraLogLog(46),
-    ThetaSketchAccumulator(47);
+    ThetaSketchAccumulator(47),
+    IntArray(48),
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:160]: Can you add more context on the code path requiring ser/de of these arrays? Based on my understanding this class is only used for ser/de the intermediate result of aggregation, so trying to understand why these changes are needed

Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:160]: This is also used by `BaseDataTableBuilder.setColumn(...)` which is used by all the agg/groupby/selection queries.

Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:160]: Array Literal is pushed down to worker in v2 engine, so need to wire it all the way.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java:160]: We do have `BaseDataTableBuilder.setColumn()` for array types, and it should not go over the general handling of `Object`. Can you double check?

Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java]: You are right, int[], long[], float[], double[] are already implemented. Just need to implement `StringArray`.

Comment (Jackie-Jiang) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java]: `String[]` is also supported natively within `DataTableBuilder`. Check the interface

File: pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java [hunk 2]
```diff
@@ -163,6 +166,37 @@ public static Expression getLiteralExpression(boolean value) {
     return expression;
   }
 
+  public static Expression getLiteralExpression(int[] value) {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/request/RequestUtils.java:169]: (nit) Move them after the single value methods

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java [hunk 3]
```diff
@@ -56,6 +56,10 @@ public void init(List<TransformFunction> arguments, Map<String, ColumnContext> c
     TransformFunction castFormatTransformFunction = arguments.get(1);
     if (castFormatTransformFunction instanceof LiteralTransformFunction) {
       String targetType = ((LiteralTransformFunction) castFormatTransformFunction).getStringLiteral().toUpperCase();
+      if ("OBJECT".equals(targetType)) {
```
Comment (Jackie-Jiang) [inline pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:59]: Hmm, what does `CAST to OBJECT` mean?

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java [hunk 4]
```diff
@@ -66,6 +66,12 @@ public static Object convert(Object value, ColumnDataType storedType) {
         if (value instanceof FloatArrayList) {
           // For ArrayAggregationFunction
           return ((FloatArrayList) value).elements();
+        } else if (value instanceof double[]) {
```
Comment (Jackie-Jiang) [inline pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java:69]: Why do we need this special handling?

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java [hunk 5]
```diff
@@ -171,13 +180,70 @@ private static RexNode evaluateLiteralOnlyFunction(RexCall rexCall, RexBuilder r
           "Caught exception while converting result value: " + resultValue + " to type: " + rexNodeType, e);
     }
     try {
+      if (rexNodeType instanceof ArraySqlType) {
+        List<Object> resultValues = new ArrayList<>();
+        assert resultValue != null;
```
Comment (Jackie-Jiang) [inline pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java:185]: Why?

File: pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java (hunk 1 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/common/ObjectSerDeUtils.java]: The issue is we didn't register the return type for v2 engine for `split` function. So it was `OBJECT` in desired return type.

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java (hunk 4 above)
Comment (xiangfu0) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java:69]: Yes, `[0.1, 0.2, 0.3]` is parsed as `FLOAT_ARRAY` but internal representation is a `double[]`.

File: pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java (hunk 3 above)
Comment (xiangfu0) [reply pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CastTransformFunction.java:59]: This happens when scalar function is not explicitly registered for the return type.

File: pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java (hunk 5 above)
Comment (xiangfu0) [reply pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotEvaluateLiteralRule.java:206]: moved the null check to above.

File: pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java [hunk 6]
```diff
@@ -1512,8 +1521,28 @@ private void computeResultsForLiteral(Literal literal, List<String> columnNames,
...
+      case DOUBLE_ARRAY_VALUE:
+        columnTypes.add(DataSchema.ColumnDataType.DOUBLE_ARRAY);
+        row.add(literal.getDoubleArrayValue());
+        break;
+      case STRING_ARRAY_VALUE:
+        columnTypes.add(DataSchema.ColumnDataType.STRING_ARRAY);
+        row.add(literal.getStringArrayValue());
+        break;
+      default:
+        throw new IllegalStateException("No able to compute results for literal - " + literal);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseBrokerRequestHandler.java:1545]:
```suggestion
        throw new IllegalStateException("Unsupported literal: " + literal);
```

File: pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/utils/TypeUtils.java:69]: Add some comment explaining this

**Review by Jackie-Jiang — APPROVED** [review]
